File: nhance/nhance/report/document_review_template/document_review_template.js
// Copyright (c) 2016, Epoch and contributors
// For license information, please see license.txt
/* eslint-disable */
var docIdss = "";
var doctype = "";
frappe.query_reports["Document Review Template"] = {
    "filters": [{
            "fieldname": "document_type",
            "label": __("Doument Type"),
            "fieldtype": "Select",
            "options": ["Sales Order", "Purchase Order", "Purchase Receipt"],
            "reqd": 1,
            "on_change": function(query_report) {
                frappe.query_report.refresh();
                frappe.query_report.set_filter_value("docIds", []);
                console.log("on_change....of for");
                frappe.query_reports["Document Review Template"].filters[1].options = "";
                frappe.query_reports["Document Review Template"].filters[1].refresh;
                frappe.query_report.refresh();

                var docName = frappe.query_report.get_filter_value("document_type");
                var docIds_filter = frappe.query_report.get_filter_value("docIds");

                frappe.query_reports["Document Review Template"].filters[1].options = docName;
                console.log("on_change....of for docName" + docName);
                frappe.query_report.refresh();
            } //end of on_change function..
        },
        {
            "fieldname": "docIds",
            "label": __("Doc Ids"),
            "fieldtype": "Link"
        },
        {
            "fieldname": "required_for_review",
            "label": __("Required for Review"),
            "fieldtype": "Select",
            "options": ["Yes", "No"],
            "default": "No",
            "on_change": function(query_report) {
                frappe.query_report.refresh();
                frappe.query_reports["Document Review Template"].filters[1].refresh;
                console.log("-----------" + frappe.query_reports["Document Review Template"].filters[2].options);
                var docName = frappe.query_report.get_filter_value("required_for_review");
                if (docName == "No") {
                    console.log("selected No options");
                    frappe.query_reports["Document Review Template"].filters[3].options = "";
                    frappe.query_reports["Document Review Template"].filters[3].refresh;
                    frappe.query_reports["Document Review Template"].filters[4].refresh;
                    // frappe.query_reports["Document Review Template"].filters[4].options = "";
                    frappe.query_reports["Document Review Template"].filters[4].read_only = 1;

                } else {
                    frappe.query_reports["Document Review Template"].filters[3].options = "Sales Invoice";
                    frappe.query_reports["Document Review Template"].filters[3].refresh;
                    frappe.query_reports["Document Review Template"].filters[4].refresh;
                    var list = ["Information/Read Only", "Accept/Reject/Enter New Value"];


                }
                frappe.query_report.refresh();
            }
        },
        {
            "fieldname": "reviewed_by",
            "label": __(" Reviewed By"),
            "fieldtype": "Link"

        },
        {
            "fieldname": "review_outcome",
            "label": __("Review Outcome"),
            "fieldtype": "Select",
            "options": ["Information/Read Only", "Accept/Reject/Enter New Value"],
            "default": "Information/Read Only",
            "on_change": function(query_report) {
                var required_for_review = frappe.query_report.get_filter_value("required_for_review");
                var docName = frappe.query_report.get_filter_value("review_outcome");
                docIdss = frappe.query_report.get_filter_value("docIds");
                doctype = frappe.query_report.get_filter_value("document_type");
                var date = function_for_date(doctype, docIdss);
                var delivery_date = JSON.stringify(date);
                console.log("date-------" + delivery_date);
                //var date1 =date.toDateString("dd-mm-yyyy");
                //console.log(date1);
                if (docName == "Accept/Reject/Enter New Value" && required_for_review == "Yes") {
                    var d = new frappe.ui.Dialog({
                        'fields': [{
                                "fieldtype": "Button",
                                "label": "Accept",
                                "fieldname": "butn"
                            },
                            {
                                "fieldtype": "Column Break",
                                "fieldname": "column_break1"
                            },
                            {
                                "fieldtype": "Button",
                                "label": "Reject",
                                "fieldname": "butn1"
                            },
                            {
                                "fieldtype": "Column Break",
                                "fieldname": "column_break2"
                            },
                            {
                                "fieldtype": "Button",
                                "label": "Enter New Value",
                                "fieldname": "butn2"
                            },
                            {
                                "fieldname": "sb_14",
                                "fieldtype": "Section Break",
                            },
                            {
                                "fieldtype": "Date",
                                "label": "Enter Date",
                                "fieldname": "date",
                                'default': delivery_date
                            },
                        ],

                    });
                    d.fields_dict.butn.$wrapper.on('click', function() {
                       // var date = $("input[data-fieldname='date']").val();
                        show_alert("clicked");
                        //selectStudentIdFromStudentDocType(batch_name);
                        d.hide();
                    });
                    d.fields_dict.butn1.$wrapper.on('click', function() {
                        show_alert('hello2');
                        console.log("doc ids--------" + docIdss);
                        function_doctype(doctype, docIdss);
                        d.hide();
                    });
                    //  d.fields_dict.butn2.$wrapper.on('click', function() {
                    //   show_alert('hello3');
                    d.fields_dict.butn2.input.onclick = function() {
                        var input_date = "";
			console.log("before get date-----"+input_date);
                        input_date = $("input[data-fieldname='date']").val();
			console.log("after get date------"+input_date);
                        var newdate = input_date.split("-").reverse().join("-");
                        show_alert(newdate);
                        var new_id =  function_doctype_cancell_amended(doctype, docIdss, newdate);
			show_alert(new_id);
			frappe.query_report.refresh();
			frappe.query_report.set_filter_value("docIds", "");
			frappe.query_report.set_filter_value("docIds", new_id);
			frappe.query_report.refresh();
			
			d.hide();
			
                       
                    }

                    //  });
                    d.show();
                } else {
                    show_alert("information of document");
                }
            }
        },
        {
            "fieldname": "details",
            "label": __("Doctype Details"),
            "fieldtype": "Select",
            "options": ["Parent", "Items", "Taxes and Charges"],
            "default": "Parent"
        }
    ]
}

function function_doctype(doctype, docIdss) {
    var expense_account = "";
    frappe.call({
        method: "nhance.nhance.report.document_review_template.document_review_template.cancel_doc",
        args: {
            "doctype": doctype,
            "docIdss": docIdss
        },
        async: false,

        callback: function(r) {
            console.log("r.message ------" + JSON.stringify(r.message));

        }
    });

}

function function_doctype_cancell_amended(doctype, docIdss, newdate) {
    var new_doctype = [];
    frappe.call({
        method: "nhance.nhance.report.document_review_template.document_review_template.cancel_and_amend_doc",
        args: {
            "doctype": doctype,
            "docIdss": docIdss,
            "newdate": newdate
        },
        async: false,

        callback: function(r) {
            console.log("r.message ------" + JSON.stringify(r.message));
	   new_doctype = r.message;
        }
    });
	return new_doctype
}

function function_for_date(doctype, docIdss) {
    var expense_account = "";
    frappe.call({
        method: "nhance.nhance.report.document_review_template.document_review_template.date_details",
        args: {
            "doctype": doctype,
            "docIdss": docIdss
        },
        async: false,

        callback: function(r) {
            console.log("r.message ------" + JSON.stringify(r.message));
            expense_account = r.message;
            console.log("expense_account---------" + expense_account);

        }
    });
    return expense_account;
}

File: nhance/nhance/report/document_review_template/document_review_template.py
# Copyright (c) 2013, Epoch and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from datetime import datetime
import datetime
from datetime import date, timedelta
import calendar
from frappe import _, msgprint

def execute(filters=None):
	global columns
	global data
	columns, data = [], []
	month_filter = filters.get("document_type")
	doctype = filters.get("document_type")
	docIds = filters.get("docIds")
	doc_filter = filters.get("details")
	#print "doctype----",doctype
	#print "docids-----",docIds
	review_outcome = filters.get("review_outcome")
	if review_outcome == "Information/Read Only":
		if doctype is not None and docIds is not None:
			doc_details = frappe.get_doc(doctype, docIds)
			if doctype == "Sales Order":
				if doc_filter == "Parent":
					columns=get_columns_sales()
					name = doc_details.name
					supplier_name = doc_details.customer_name
					delivery_date = doc_details.delivery_date
					doc_type = doc_details.order_type
					creation_date = doc_details.transaction_date
					company = doc_details.company
					total_qty = doc_details.total_qty
					grand_total = doc_details.grand_total
					docstatus = doc_details.status
					data.append([name , supplier_name,doc_type,company,creation_date,creation_date,total_qty,
					grand_total,docstatus])
					
				elif doc_filter == "Items":
					columns=get_columns_sales_items()
					#doc_details = frappe.get_doc(doctype, docIds)
					sales_items = doc_details.items
					for items in sales_items:
						item_code = items.item_code
						item_name = items.item_name
						item_qty = items.qty
						stock_uom = items.stock_uom
						delivery_warehouse = items.warehouse
						rate = items.net_rate
						amount = items.net_amount
						delivery_date = items.delivery_date
						data.append([item_code,item_name,item_qty,stock_uom,rate,
						amount,delivery_warehouse,delivery_date])
				elif doc_filter == "Taxes and Charges":
					columns=get_columns_sales_items()
					#doc_details = frappe.get_doc(doctype, docIds)
					sales_taxes = doc_details.taxes
					if sales_taxes is not None:
						for taxes in sales_taxes:
							account_head = items.account_head
							charge_type = items.charge_type
							tax_amount = items.tax_amount
							rate = items.net_rate
							cost_center = items.cost_center
							data.append([account_head,charge_type,tax_amount,rate,
							cost_center])
			if doctype == "Purchase Order":
				if doc_filter == "Parent":
					columns=get_columns_purchase()
					#name =  frappe.db.sql("""select name from `tabSales Order` where amended_from = %s""",(docIds))
					name = doc_details.name
					supplier_name = doc_details.supplier_name
					schedule_date = doc_details.schedule_date
					#doc_type = doc_details.order_type
					transaction_date = doc_details.transaction_date
					company = doc_details.company
					docstatus = doc_details.status
					total_qty = doc_details.total_qty
					grand_total = doc_details.grand_total
					data.append([name , supplier_name,company,transaction_date,schedule_date,total_qty,
					grand_total,docstatus])
					
				elif doc_filter == "Items":
					columns=get_columns_purchase_items()
					#doc_details = frappe.get_doc(doctype, docIds)
					sales_items = doc_details.items
					for items in sales_items:
						item_code = items.item_code
						item_name = items.item_name
						item_qty = items.qty
						stock_uom = items.stock_uom
						delivery_warehouse = items.warehouse
						rate = items.net_rate
						amount = items.net_amount
						delivery_date = items.schedule_date
						data.append([item_code,item_name,item_qty,stock_uom,rate,
						amount,delivery_warehouse,delivery_date])
				elif doc_filter == "Taxes and Charges":
					columns=get_columns_sales_items()
					#doc_details = frappe.get_doc(doctype, docIds)
					sales_taxes = doc_details.taxes
					if sales_taxes is not None:
						for taxes in sales_taxes:
							account_head = items.account_head
							charge_type = items.charge_type
							tax_amount = items.tax_amount
							rate = items.net_rate
							cost_center = items.cost_center
							data.append([account_head,charge_type,tax_amount,rate,
							cost_center])
	elif review_outcome == "Accept/Reject/Enter New Value":
		if doctype is not None and docIds is not None:
			doc_details = frappe.get_doc(doctype, docIds)
			if doctype == "Sales Order":
				if doc_filter == "Parent":
					columns=get_columns_sales()
					#name =  frappe.db.sql("""select name from `tabSales Order` where amended_from = %s""",(docIds))
					name = doc_details.name
					supplier_name = doc_details.customer_name
					delivery_date = doc_details.delivery_date
					doc_type = doc_details.order_type
					creation_date = doc_details.transaction_date
					company = doc_details.company
					docstatus = doc_details.status
					total_qty = doc_details.total_qty
					grand_total = doc_details.grand_total
					data.append([name , supplier_name,doc_type,company,creation_date,delivery_date,total_qty,
					grand_total,docstatus])
					
				elif doc_filter == "Items":
					columns=get_columns_sales_items()
					#doc_details = frappe.get_doc(doctype, docIds)
					sales_items = doc_details.items
					for items in sales_items:
						item_code = items.item_code
						item_name = items.item_name
						item_qty = items.qty
						stock_uom = items.stock_uom
						delivery_warehouse = items.warehouse
						rate = items.net_rate
						amount = items.net_amount
						delivery_date = items.delivery_date
						data.append([item_code,item_name,item_qty,stock_uom,rate,
						amount,delivery_warehouse,delivery_date])
				elif doc_filter == "Taxes and Charges":
					columns=get_columns_sales_items()
					#doc_details = frappe.get_doc(doctype, docIds)
					sales_taxes = doc_details.taxes
					if sales_taxes is not None:
						for taxes in sales_taxes:
							account_head = items.account_head
							charge_type = items.charge_type
							tax_amount = items.tax_amount
							rate = items.net_rate
							cost_center = items.cost_center
							data.append([account_head,charge_type,tax_amount,rate,
							cost_center])
			if doctype == "Purchase Order":
				if doc_filter == "Parent":
					columns=get_columns_purchase()
					#name =  frappe.db.sql("""select name from `tabSales Order` where amended_from = %s""",(docIds))
					name = doc_details.name
					supplier_name = doc_details.supplier_name
					schedule_date = doc_details.schedule_date
					#doc_type = doc_details.order_type
					transaction_date = doc_details.transaction_date
					company = doc_details.company
					docstatus = doc_details.status
					total_qty = doc_details.total_qty
					grand_total = doc_details.grand_total
					data.append([name , supplier_name,company,transaction_date,schedule_date,total_qty,
					grand_total,docstatus])
					
				elif doc_filter == "Items":
					columns=get_columns_purchase_items()
					#doc_details = frappe.get_doc(doctype, docIds)
					sales_items = doc_details.items
					for items in sales_items:
						item_code = items.item_code
						item_name = items.item_name
						item_qty = items.qty
						stock_uom = items.stock_uom
						delivery_warehouse = items.warehouse
						rate = items.net_rate
						amount = items.net_amount
						delivery_date = items.schedule_date
						data.append([item_code,item_name,item_qty,stock_uom,rate,
						amount,delivery_warehouse,delivery_date])
				elif doc_filter == "Taxes and Charges":
					columns=get_columns_sales_items()
					#doc_details = frappe.get_doc(doctype, docIds)
					sales_taxes = doc_details.taxes
					if sales_taxes is not None:
						for taxes in sales_taxes:
							account_head = items.account_head
							charge_type = items.charge_type
							tax_amount = items.tax_amount
							rate = items.net_rate
							cost_center = items.cost_center
							data.append([account_head,charge_type,tax_amount,rate,
							cost_center])
	return columns, data



def get_columns_sales():
	return [
		_("Sales Order Number") + "::150", 
		_("Customer name ") + "::180",
		_("Order Type") + "::120",
		_("Company")+"::130",
		_("Creation Date") + "::150", 
		_("Delivery Date") + "::180",
		_("Total Quantity") + "::120",
		_("Grand Total (INR)")+"::130",
		_("Status")+"::130"
		
	]
def get_columns_purchase():
	return [
		_("Sales Order Number") + "::150", 
		_("Customer name ") + "::180",
		_("Company")+"::130",
		_("Creation Date") + "::150", 
		_("Schedule Date") + "::180",
		_("Total Quantity") + "::120",
		_("Grand Total (INR)")+"::130",
		_("Status")+"::130"
		
	]
def get_columns_sales_items():
	return [
		_("Item Code") + "::150", 
		_("Item Name") + "::180",
		_("Quantity") + "::120",
		_("Stock UOM")+"::130",
		_("Rate (INR)") + "::150", 
		_("Amount (INR)") + "::180",
		_("Delivery Warehouse") + "::120",
		_("Delivery Date")+"::130"
		
	]
def get_columns_purchase_items():
	return [
		_("Item Code") + "::150", 
		_("Item Name") + "::180",
		_("Quantity") + "::120",
		_("Stock UOM")+"::130",
		_("Rate (INR)") + "::150", 
		_("Amount (INR)") + "::180",
		_("Warehouse") + "::120",
		_("Schedule Date")+"::130"
		
	]
def get_columns_sales_taxes():
	return [
		_("Account Head") + "::150", 
		_("Charge Type") + "::180",
		_("Tax Amount") + "::120",
		_("Rate")+"::130",
		_("Cost Center") + "::150"
		
	]
@frappe.whitelist()
def cancel_doc(doctype,docIdss):
	stock_requisition = frappe.get_doc(doctype, docIdss)
	stock_requisition.cancel()
	frappe.msgprint("The doctype "+doctype +" and docId "+docIdss+" is cancelled successfully!!")
	return 1
@frappe.whitelist()
def date_details(doctype,docIdss):
	stock_requisition = ""
	if doctype == "Sales Order":
		stock_requisition =  frappe.db.get_value(doctype,docIdss,'delivery_date')
		#stock_requisition = frappe.get_doc(doctype, docIdss,"delivery_date")
		#date = stock_requisition.delivery_date
		#date = date.strftime('%d-%m-%Y')
		#print "stock_requisition-----------",stock_requisition.delivery_date
	elif doctype == "Purchase Order":
		stock_requisition =  frappe.db.get_value(doctype,docIdss,'schedule_date')
	return stock_requisition

@frappe.whitelist()
def cancel_and_amend_doc(doctype,docIdss,newdate):
	stock_requisition = frappe.get_doc(doctype, docIdss)
	#print "stock_requisition------------",stock_requisition
	#stock_requisition.cancel()
	#date = str(date).strftime('%Y-%m-%d')
	#print "date-----",newdate
	#date = datetime(date)
	#date = datetime.datetime.strptime(date,"%Y-%m-%d")
	#stock_requisition.delivery_date = date
	
	stock_requisition.cancel() 
	frappe.msgprint("The doctype "+doctype +" and docId "+docIdss+" is cancelled successfully!!")
	new_pr = frappe.copy_doc(stock_requisition) 
	new_pr.amended_from = stock_requisition.name 
	new_pr.status = "Draft" 
	new_pr.insert()
	#new_pr.delivery_date = ""
	#new_pr.delivery_date = newdate
	#print "new_doc--------",new_pr.delivery_date
	new_pr.save()
	sales_name = new_pr.name
	#print "name------------",sales_name
	new_pr.submit()
	frappe.msgprint(" Successfully created new Doc "+doctype+" as Amended "+sales_name+" !!")
	if doctype == "Sales Order":
		frappe.db.sql("""UPDATE `tabSales Order` SET delivery_date = '"""+str(newdate)+"""' where name = '"""+sales_name+"""'""")
		frappe.db.sql("""UPDATE `tabSales Order Item` SET delivery_date = '"""+str(newdate)+"""' where parent = '"""+sales_name+"""'""")
	elif doctype == "Purchase Order":
		frappe.db.sql("""UPDATE `tabPurchase Order` SET schedule_date = '"""+str(newdate)+"""' where name = '"""+sales_name+"""'""")
		frappe.db.sql("""UPDATE `tabPurchase Order Item` SET schedule_date = '"""+str(newdate)+"""' where parent = '"""+sales_name+"""'
		""")
	
	return sales_name

